pkg/registry/apis/userstorage/strategy.go

```diff
@@ -0,0 +1,109 @@
+package userstorage
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/grafana/grafana/pkg/apimachinery/identity"
+	"github.com/grafana/grafana/pkg/apimachinery/utils"
+	grafanaregistry "github.com/grafana/grafana/pkg/apiserver/registry/generic"
+	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/apiserver/pkg/registry/rest"
+)
+
+var (
+	// Target for user storage size < 3MB
+	userstorageSize = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "userstorage",
+		Name:      "size",
+		Help:      "The current size in bytes of the user storage for a service",
+	}, []string{"name"})
+)
+
+type genericStrategy interface {
+	rest.RESTCreateStrategy
+	rest.RESTUpdateStrategy
+}
+
+type userstorageStrategy struct {
+	genericStrategy
+
+	registerer prometheus.Registerer
+}
+
+var once sync.Once
+
+func newStrategy(typer runtime.ObjectTyper, gv schema.GroupVersion, registerer prometheus.Registerer) *userstorageStrategy {
+	once.Do(func() {
+		if registerer != nil {
+			registerer.MustRegister(
+				userstorageSize,
+			)
+		}
+	})
+	genericStrategy := grafanaregistry.NewStrategy(typer, gv)
+	return &userstorageStrategy{genericStrategy, registerer}
+}
+
+func compareResourceNameAndUserUID(name string, u identity.Requester) bool {
+	// ObjectMeta.Name should follow the format <service>:<user_uid>
+	nameSplit := strings.Split(name, ":")
+	if len(nameSplit) != 2 {
+		return false
+	}
+	objUserUID := nameSplit[1]
+
+	// u.GetUID() returns user:<user_uid> so we need to remove the user: prefix
+	userUID := strings.Split(u.GetUID(), ":")
+	if len(userUID) != 2 {
+		return false
+	}
+
+	return objUserUID == userUID[1]
+}
+
+func registerSize(obj runtime.Object) {
+	meta, err := utils.MetaAccessor(obj)
+	if err != nil {
+		return
+	}
+
+	b := new(bytes.Buffer)
+	if err := json.NewEncoder(b).Encode(obj); err != nil {
+		return
+	}
+	userstorageSize.WithLabelValues(meta.GetName()).Set(float64(b.Len()))
```

Comment:
Could we change this to avoid incorporating the UID in the label, since this will lead to cardinality explosion?

We could certainly measure this by service instead:
```suggestion
	svc := strings.Split(meta.GetName(), ":")[1]
	userstorageSize.WithLabelValues(svc).Set(float64(b.Len()))
```

Reply:
doesn't that defeat [the original purpose](https://github.com/grafana/grafana/pull/95557#discussion_r1832884764) of this metric? If I understand correctly, the worry is that we run into entries too large (reaching the 3MB limit). If we don't track the UID, we don't know if some service is overusing the feature.

Reply:
Good point, sorry – Monday's been a bit hectic for me. Still, the problem with cardinality explosion can be real, so we need to reign this in somehow.

How about using a histogram instead of a gauge? I think this should give us enough resolution.

Something like this maybe:
```diff
diff --git a/pkg/registry/apis/userstorage/strategy.go b/pkg/registry/apis/userstorage/strategy.go
index 020b62c31f7..62d05c0391a 100644
--- a/pkg/registry/apis/userstorage/strategy.go
+++ b/pkg/registry/apis/userstorage/strategy.go
@@ -4,6 +4,7 @@ import (
 	"bytes"
 	"context"
 	"encoding/json"
+	"errors"
 	"fmt"
 	"strings"
 	"sync"
@@ -20,11 +21,12 @@ import (
 
 var (
 	// Target for user storage size < 3MB
-	userstorageSize = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+	userstorageSize = prometheus.NewHistogramVec(prometheus.HistogramOpts{
 		Namespace: "userstorage",
-		Name:      "size",
-		Help:      "The current size in bytes of the user storage for a service",
-	}, []string{"name"})
+		Name:      "object_size_bytes",
+		Help:      "Histogram of user storage object sizes in bytes, broken down by service name",
+		Buckets:   prometheus.ExponentialBucketsRange(1024, 8*1024*1024, 8), // From 1 KB to 8 MB
+	}, []string{"service"})
 )
 
 type genericStrategy interface {
@@ -53,12 +55,10 @@ func newStrategy(typer runtime.ObjectTyper, gv schema.GroupVersion, registerer p
 }
 
 func compareResourceNameAndUserUID(name string, u identity.Requester) bool {
-	// ObjectMeta.Name should follow the format <service>:<user_uid>
-	nameSplit := strings.Split(name, ":")
-	if len(nameSplit) != 2 {
+	var parsedName storageObjectName
+	if err := parseName(name, &parsedName); err != nil {
 		return false
 	}
-	objUserUID := nameSplit[1]
 
 	// u.GetUID() returns user:<user_uid> so we need to remove the user: prefix
 	userUID := strings.Split(u.GetUID(), ":")
@@ -66,7 +66,7 @@ func compareResourceNameAndUserUID(name string, u identity.Requester) bool {
 		return false
 	}
 
-	return objUserUID == userUID[1]
+	return parsedName.UID == userUID[1]
 }
 
 func registerSize(obj runtime.Object) {
@@ -75,11 +75,17 @@ func registerSize(obj runtime.Object) {
 		return
 	}
 
+	var parsedName storageObjectName
+	if err := parseName(meta.GetName(), &parsedName); err != nil {
+		return
+	}
+
 	b := new(bytes.Buffer)
 	if err := json.NewEncoder(b).Encode(obj); err != nil {
 		return
 	}
-	userstorageSize.WithLabelValues(meta.GetName()).Set(float64(b.Len()))
+
+	userstorageSize.WithLabelValues(parsedName.Service).Observe(float64(b.Len()))
 }
 
 // Validate ensures that when creating a userstorage object, the name matches the user id.
@@ -107,3 +113,21 @@ func (g *userstorageStrategy) ValidateUpdate(ctx context.Context, obj, old runti
 	registerSize(obj)
 	return field.ErrorList{}
 }
+
+type storageObjectName struct {
+	Service string
+	UID     string
+}
+
+func parseName(name string, dst *storageObjectName) error {
+	// ObjectMeta.Name should follow the format <service>:<user_uid>
+	vals := strings.Split(name, ":")
+	if len(vals) != 2 {
+		return errors.New("invalid name")
+	}
+
+	dst.Service = vals[0]
+	dst.UID = vals[1]
+
+	return nil
+}
```